### PR TITLE
kernel: support dynamic thread stack allocation

### LINKED ISSUE #26999: Add a Stack Allocation API
Relates to #24714

We want an API such that we can pass in a desired stack size, and a suitably aligned pointer to a stack object will be returned, with all reserved space, alignment, and permissions considerations taken care of.

### PR DESCRIPTION
**Pull request by cfriedt**
Opening this up for review. It's a biggie, and has some wide-reaching implications.

TL;DR - we should be able to support automatic pthread stack allocation with this. It's in the Roadmap (#51211).

However, at least with GCC, this does a lot more though. With compliant POSIX threads `<pthread.h>`, we also get

* ISO C11 `<threads.h>`
* ISO C++11 `<thread>` (`<mutex>`, `<condition_variable>`, `<semaphore>`), etc.

Later, we can optimize-out the POSIX layer of course, which I would like to take great care in doing with Keith and Stephanos.

Taking a step back, excluding coroutines, that represents the de-facto threading model for approximately the last 25 years, and probably the de-facto threading model for most other programming languages for the foreseeable future.

* MicroPython threads
* Rust threads
* Lua threads
* Zig threads
* ...

The API supports both safety-critical and non-safety critical because it uses the same interface for both pool-based and heap-based thread stack allocations.

Covered in my talks at EOSS 2023.

Thanks for the help everyone 🙏 ❤️ 
@andrewboie , @pfalcon, @andyross , @peter-mitsis , @nashif , @dcpleung , @ceolin , @stephanosio , @keith-packard ... I really could keep going.

Fixes #26999

There is some icing on the top added by @ceolin for `k_object` management in #59773

**BELOW FOR POSTERITY**

Just a WIP at the moment. Working fine for cortex-m3.

```shell
$ west build -p auto -b qemu_cortex_m3 -t run tests/kernel/threads/dynamic_thread_stack
...
*** Booting Zephyr OS build zephyr-v3.0.0-2000-g3ccf6cd86f23  ***
Running test suite dynamic_thread_stack
===================================================================
START - test_dynamic_thread_stack
Hello, dynamic world!
 PASS - test_dynamic_thread_stack in 0.1 seconds
===================================================================
Test suite dynamic_thread_stack succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

### CODE REVIEW

**Comment by cfriedt** [comment]
@dcpleung - would it make sense to separate the redirection layer into its own PR?

**Comment by dcpleung** [comment]
Architectures have `arch_mem_map()` and `arch_mem_unmap()`. Adding a bunch of `arch_mm_drv_*()` is going to confuse developers. Maybe add a system mm driver to utilize `arch_mem_*()` instead?

**Comment by cfriedt** [comment]
> Architectures have `arch_mem_map()` and `arch_mem_unmap()`. Adding a bunch of `arch_mm_drv_*()` is going to confuse developers. Maybe add a system mm driver to utilize `arch_mem_*()` instead?

Done-ish? I didn't bother creating a sys interface, it seemed to be overkill for now.

**Comment by kestewart** [comment]
Can the formal requirements that this new capability provides be listed as part of this pull request?

**Comment by cfriedt** [comment]
> Can the formal requirements that this new capability provides be listed as part of this pull request?  

@kestewart  I can certainly try - do we have a template for something like that?

**Comment by romkell** [comment]
Triggered by the Safety Committee meeting I just had a look at the code.
I am glad you already considered safety. The discussion in the meeting was if dynamic stacks should not be supported by safety configuring it away in the middle of the code is manageable for safety with the increasing number of such features in the long run.

I have never written an RTOS, only used different. But would it be feasible to do a layered approach where the dynamic part - allocating / freeing stack (and thread context (?)) - in a e.g. dyn_thread.c which then passes those objects to the existing static version?
Is it required that we have the same API for both?
If it is required, could that layer on top not just route the function call through to the current static implementation without doing anything when dynamic stack is configured off?
Why is only the stack dynamic and not all objects (hence the thread struct too and a handle be returned to the top most object?

With different API:
> 
    -------------------------------------------
    |         dyn_thread.c                    |
    | (alloc / free stack and thread context) |
    -------------------------------------------
    |         thread.c                        |
    -------------------------------------------

With same API (route through without doing ideally nothing, when dynamic stacks is disabled. thread.c could even be two files, thread_w_dyn.c and thread_wo_dyn.c.
> 
    -------------------------------------------
    |         thread.c                        |
    | (alloc / free stack and thread context  |
    |  if dynamic stacks is enabled)          |
    -------------------------------------------
    |         static_thread.c                 |
    -------------------------------------------

It is a more elaborated version (if feasible at all), which likely might also consume more FLASH and RAM.

**Comment by cfriedt** [comment]
> Triggered by the Safety Committee meeting I just had a look at the code. I am glad you already considered safety. The discussion in the meeting was if features, such as dynamic stacks, is manageable for safety with the increasing number of such features in the long run, if it is done by code fragments being configured (#fidef-d) away in different locations in the a source file.

That's a good question. The emphasis I would suggest is to focus less on the actual code and more on using Kconfig to configure software based on policy and CI to test that policy.

Code can always be changed, but the policy is what is most important, and finding a "language" to verify that policy is key to making it manageable.

 I've really only added this Kconfig option as an illustration of how some kind of policy could be done (after all, Kconfig is for configuring software).
 
 Another unexplored facet of safety-critical is that userspace *should* be allowed to do things that would normally not be considered OK in kernel space in a safety critical system. Userspace should be allowed to crash all over the place. In fact, that in itself, would be an excellent way to test that kernel space is safe.

> I have never written an RTOS, only used different. But would it be feasible to do a layered approach where the dynamic part - allocating / freeing stack (and thread context (?)) - in a e.g. dyn_thread.c which then passes those objects to the
> 
> ```
> -------------------------------------------
> |         dyn_thread.c                    |
> | (alloc / free stack and thread context) |
> -------------------------------------------
> |         thread.c                        |
> -------------------------------------------
> ```

That's a great suggestion. 

> With same API (route through without doing ideally nothing, when dynamic stacks is disabled. thread.c could even be two files, thread_w_dyn.c and thread_wo_dyn.c.


> ```
> -------------------------------------------
> |         thread.c                        |
> | (alloc / free stack and thread context  |
> |  if dynamic stacks is enabled)          |
> -------------------------------------------
> |         static_thread.c                 |
> -------------------------------------------
> ```
> 
> It is very likely a more elaborated version (if feasible at all), which likely might also consume more FLASH and RAM too, compared to your existing approach.

Also a great suggestion.

I don't anticipate that I will have a lot of time to implement either of them unfortunately, as I've been at this particular fix already for over 2 years (mostly in my spare time) and have only just barely reached a solution, and for the most part my use cases are non-safety-critical.

However, I think what would probably help the safety-critical use case is to
* explicitly define policies for inclusion / exclusion of safety-critical features
  * this is a big task, but is ultimately finite in size
  * devise tests that can be run in CI to check those policies
  * when it comes to safety, I always say the testsuite is 10x more valuable than the actual code
  * generate coverage reports
* add documentation about reusable design patterns such as the two you mentioned above
  * copying and modifying a template is a scalable solution for developers

**Comment by cfriedt** [comment]
> > Architectures have `arch_mem_map()` and `arch_mem_unmap()`. Adding a bunch of `arch_mm_drv_*()` is going to confuse developers. Maybe add a system mm driver to utilize `arch_mem_*()` instead?
> 
> I didn't bother creating a sys interface, it seemed to be overkill for now. 

@dcpleung - definitely not overkill now that there is a clearer picture of where a sys interface is needed.

* x86 gaurd pages / random stuff
* arm MPU (we knew this would be painful)
* RiscV PMP and odd quirks

With more esoteric architectures like cavs, xtensa, etc, I am even less sure, bit the interface would certainly help.

File: tests/kernel/threads/dynamic_thread_stack/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+common:
+    tags: kernel security
+    min_ram: 16
+    # ARM MPU is a bit tricky
+    filter: not CONFIG_ARM_MPU
+    # oddly both, qemu_x86 and qemu_riscv work fine
+    platform_exclude: qemu_x86_nommu qemu_x86_virt qemu_x86_tiny
```
Comment (nashif) [inline tests/kernel/threads/dynamic_thread_stack/testcase.yaml:7]: do not like this level of filtering. Also think about running this without --integration, which will be the case when this lands in main branch, we need to make sure it does not break anything beyond those.

Comment (cfriedt) [reply tests/kernel/threads/dynamic_thread_stack/testcase.yaml:7]: Agreed - ideally it would be possible to drill down and figure out what is going on in some of these filtered platforms. I'm not familiar enough with all of them to say what the issues are with certainty

**Comment by peter-mitsis** [comment]
I'm going to chime in and add my own two cents.  Overall, I really like the proposal.

I will second romkell's proposal about having a wrapper routine for the dynamic memory allocation and freeing. Such an approach has an additional benefit--it can allow for different allocation/freeing strategies. This can be of benefit to safety profiles that may not be permitted to use malloc/free as their behavior can not be reliably predicted. Such profiles could obtain stack memory from a pool of fixed size buffers, and return the buffer to the pool when done--a well established and allowed practice in safety systems. That would allow most of  the benefits of dynamic stack allocation in a certifiied/safety based system.

File: tests/kernel/threads/dynamic_thread_stack/testcase.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      qemu_riscv64_smp qemu_malta qemu_malta_be qemu_leon3 qemu_arc_hs
+      qemu_arc_hs6x
+    integration_platforms:
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply tests/kernel/threads/dynamic_thread_stack/testcase.yaml:10]: Most likely, I'll augment the existing `kernel/threads/thread_stack` test. A separate & simpler test was just more useful when developing the feature.

File: tests/kernel/threads/dynamic_thread_stack/testcase.yaml (hunk 1 above)
Comment (cfriedt) [reply tests/kernel/threads/dynamic_thread_stack/testcase.yaml:7]: Most likely, I'll augment the existing `kernel/threads/thread_stack` test. A separate & simpler test was just more useful when developing the feature.

**Comment by dcpleung** [comment]
Took a quick look at the sys_mm to arch_mm interface:

* `sys_mm_drv_map_page(virt, phys, sys_mm_flags)` can call `arch_mem_map(virt, phys, CONFIG_MM_DRV_PAGE_SIZE, arch_mm_flags)` with a translation of flags from `SYS_MM_MEM_CACHE_*` and `SYS_MM_MEM_PERM_*` to `K_MEM_CACHE_*` and `K_MEM_PERM_*`.
* `sys_mm_drv_unmap_page(virt)` can call `arch_mem_unmap(virt, CONFIG_MM_DRV_PAGE_SIZE)` directly.
* `sys_mm_drv_page_phys_get(virt, phys)` can call `arch_page_phys_get(virt, phys)` directly.
* Others `sys_mm_drv_*()` have common implementations in `drivers/mm/mm_drv_common.c`. So no need to worry about them.

(Hopefully I didn't miss anything.)

**Comment by cfriedt** [comment]
@andyross - what are your thoughts here in terms of safety? I think the game plan from a few weeks ago was to add dynamic allocation in `k_thread_create()` but I guess it's not a great option from the safety critical perspective.

https://github.com/zephyrproject-rtos/zephyr/pull/43730#issuecomment-1071004966

It's not exactly a simple wrapper either mainly because the majority of this needs to be done in kernel context rather than user context (so with all of the `z_verify` / `z_impl` machinery).

The other issue is that the stack teardown is a bit tricky
* `CONFIG_ARCH_HAS_THREAD_ABORT` currently turns `z_thread_abort()` into a syscall *sometimes*
* maybe just for aarch32 / posix? What aarch32 is doing here is also a bit weird and could probably be a bit more linear

**Review by andyross — COMMENTED** [review]
Saw this come by and realized I still hadn't reviewed it.  Some nitpicks and notes.  Looks pretty solid to me.

**Comment by cfriedt** [comment]
Thanks for your reviews @andyross - will rework when I get back to the great white north

File: kernel/Kconfig [hunk 3]
```diff
@@ -9,6 +9,14 @@ module = KERNEL
 module-str = kernel
 source "subsys/logging/Kconfig.template.log_config"
 
+config SAFETY_CRITICAL
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply kernel/Kconfig:12]: Done!

File: include/zephyr/kernel/thread.h [hunk 4]
```diff
@@ -54,7 +54,7 @@ struct _thread_base {
 	_wait_q_t *pended_on;
 
 	/* user facing 'thread options'; values defined in include/kernel.h */
-	uint8_t user_options;
+	uint16_t user_options;
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply include/zephyr/kernel/thread.h:57]: Done!

File: include/zephyr/kernel/thread.h [hunk 5]
```diff
@@ -152,6 +152,11 @@ struct _thread_stack_info {
 	 * is the initial stack pointer for a thread. May be 0.
 	 */
 	size_t delta;
+
+#if defined(CONFIG_DYNAMIC_THREAD_STACKS)
+	void *stack_obj_start;
+	size_t stack_obj_size;
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply include/zephyr/kernel/thread.h:158]: Cool - looks to be the case. I may have been a bit hasty because I wasn't sure if a value was getting adjusted somewhere.

File: kernel/Kconfig [hunk 6]
```diff
@@ -212,6 +212,23 @@ config THREAD_USERSPACE_LOCAL_DATA
...
+config DYNAMIC_THREAD_STACKS
+	bool "Dynamically allocated thread stacks [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	depends on !SAFETY_CRITICAL
+	depends on THREAD_STACK_INFO
+	help
+	  Enable dynamically allocated thread stacks
+
+if DYNAMIC_THREAD_STACKS
+config DYNAMIC_THREAD_STACKS_DEFAULT_SIZE
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply kernel/Kconfig:224]: Gone!

File: kernel/thread.c [hunk 7]
```diff
@@ -456,17 +456,53 @@ static char *setup_thread_stack(struct k_thread *new_thread,
...
 
+#ifdef CONFIG_DYNAMIC_THREAD_STACKS
+	if (stack == NULL && (new_thread->base.user_options & K_ALLOC)) {
+		size_t align;
+
+		if (stack_size == 0) {
+			stack_size = CONFIG_DYNAMIC_THREAD_STACKS_DEFAULT_SIZE;
+		}
+
+#ifdef CONFIG_USERSPACE
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply kernel/thread.c:476]: I ended up getting implicit-function-declaration, so left it in

**Comment by cfriedt** [comment]
> Smallish typo: gaurantee vs guarantee
> 
> I still feel "config SAFETY_CRITICAL" does not belong to kernel/Kconfig but to the highest possible KConfig file. Safety will not only affect the kernel but also other parts such as fs, logging according to the safety scope defined in https://www.zephyrproject.org/wp-content/uploads/sites/38/2022/02/Zephyr-Overview-2022Q1-Public.pdf. And in fs, it might be just NVS which can be safety certified (effort-wise or base code-wise). So even there might be a selection.

I agree, but I'm also not a part of the Safety committee. Kconfig options might also be finer grained, so if there were an option for e.g. `CONFIG_MISRA_C_2012`, it might select or deselect a number of other options like `CONFIG_DYNAMIC_THREAD_STACKS` or something along those lines.

I'll try to focus on the technical aspect of this change first though, and then tackle the policy aspect later. Preferably the safety committee would architect how policy is implemented first, and then I will just make any required changes to Kconfig, SPDX, etc.

Technical changes will likely be made in this order:
1. @romkell and @peter-mitsis 's suggestions to separate memory allocation from thread setup
2. @dcpleung's suggestions w.r.t. the MM interface
3. @nashif's concerns with testing (I think that may be tied to the MM interface as well)

**Review by andyross — DISMISSED** [review]
Can't see anything worth complaining about anymore.  No reason not to leave a +1 (even though it'll surely be clobbered).

Regarding the kconfig name: it's important to recognize that the thing being abstracted isn't "safety critical" in the sense of "This Code Is Safer", but instead "this app is intended to comply with safety certification requirements that permit only a subset of common RTOS APIs and features".

Whether MISRA_SANE is a good name or not I leave for someone else to decide.  It was a bit of whimsy on my part (one that runs afoul of modern inclusivity standards) when faced with "?!%!@ I can't use alloca()!?", and probably should be renamed.  If nothing else, MISRA isn't the only such standard.

**Comment by andyross** [comment]
> oh alloca... what could possibly go wrong??

To be fair: the specific use case was some enumeration code for the rbtree, for which the obvious fallback was a recursive implementation anyway.  So stack allocation was going to be non-constant regardless.  alloca() was just more efficient.

And FWIW: the MISRA_SANE choice was actually to size a **maximum theoretical tree depth**[1], thus "fixing" the "unconstrained" allocation by literally writing pessimal code.  Thus the cranky kconfig name.

[1] To be fair, it's a balanced tree, so even on a 64 bit platform that isn't a wildly unreasonable size, a few hundred bytes or something like that.

**Comment by cfriedt** [comment]
Minor update here.
* Separated thread stack allocation to its own API based on @romkell and @peter-mitsis suggestions
* Removed Kconfig about safety-critical-ness (not worrying about policy right now)
* Disabled `CONFIG_DYNAMIC_THREAD_ALLOC` tests for now, because there is still the open issue of the MM interface @dcpleung 

On systems without an MMU / MPU, dynamic thread stacks are working nicely.

A couple of additional notes:
* Eventually it would be good to simply pair thread stacks with threads, and turn this more into a "dynamic thread" facility rather than a "dynamic thread stack" facility
* With that, it would probably make sense to create some API macros for thread pool creation, so that that code becomes more of a library API, and there is a *system* thread pool separate from more application-specific *user* thread pools

@dcpleung @andyross - I wonder if we should plan to attend the dev-review meeting soon. I could certainly use a refresher on exactly what MM facilities I would need here. Unless it's pretty obvious to either of you and it could fit in a comment.

Basically, we just need to be able to indicate to the MMU / MPU that a region of memory should be memory protected in the 'alloc' function, and it should be memory-unprotected in the 'free' function.

**Comment by dcpleung** [comment]
There are `sys_mm_drv_update_page_flags()` and `sys_mm_drv_update_region_flags()` if you want to change flags on-the-fly.

File: tests/kernel/threads/dynamic_thread_stack/prj.conf [hunk 8]
```diff
@@ -0,0 +1,14 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_INIT_STACKS=y
+CONFIG_THREAD_STACK_INFO=y
+CONFIG_DYNAMIC_THREAD=y
+CONFIG_DYNAMIC_THREAD_POOL_SIZE=2
+# See Issue #12345 - need to enable MMU / MPU / arch interface
```
Comment (evgeniy-paltsev) [inline tests/kernel/threads/dynamic_thread_stack/prj.conf:7]: Hmm, I guess there should be actual issue number? #12345 seems to be unrelated.

**Review by evgeniy-paltsev — DISMISSED** [review]

**Comment by cfriedt** [comment]
ripped off the band-aid (i.e. enabled test permutations for userspace and heap allocation, disabled filter for `CONFIG_MMU` or `CONFIG_MPU`). 

Let's see where the troublesome spots are..

I would imagine that the most informative platforms will likely be `qemu_x86_64` and `qemu_cortex_a53`, but likely some 32-bit platforms will have something to say. When I ran this locally, ~66% of cases were passing.

**Comment by cfriedt** [comment]
My gosh, I need a vacation just to get the time to work on this 😅

**Comment by cfriedt** [comment]
Just gave this one a rebase.

**Comment by peter-mitsis** [comment]
Aside from the pesky compliance checks, this looks good to me. :)

**Comment by ceolin** [comment]
@cfriedt I am looking into it and I am able to create the kobject for an user thread and run it as supervisor, but there are still things missing to allow it run as user mode. The option to tell if a stack is user stack limits makes that stacks created without it can't change to user mode.

**Comment by cfriedt** [comment]
> The option to tell if a stack is user stack limits makes that stacks created without it can't change to user mode.

I think there is a macro for that ... something like `...IS_USER_CAPABLE()`

**Comment by ceolin** [comment]
> > The option to tell if a stack is user stack limits makes that stacks created without it can't change to user mode.
> 
> I think there is a macro for that ... something like `...IS_USER_CAPABLE()`

there is  ... the point is that normally when we define a thread stack with `K_THREAD_STACK_DEFINE` if userspace is enabled it will ensure the minimum page alignment + space for privilege escalation and any thread created with this can become a user thread. The thing that is blocking me now is that the kobject for the thread stack can't be in the same pages of the stack itself, otherwise the user thread will be able to manipulate it.

**Comment by cfriedt** [comment]
Yes - exactly! So there needs to be two dynamic allocations for the thread stack

**Comment by ceolin** [comment]
@cfriedt userspace support https://github.com/zephyrproject-rtos/zephyr/pull/59773

**Comment by cfriedt** [comment]
Flavio & I will likely merge our PRs tomorrow, so no rush at the moment.

**Comment by cfriedt** [comment]
@ceolin & I discussed it, and we opted to merge this PR first, and subsequently merge https://github.com/zephyrproject-rtos/zephyr/pull/59773 afterward.

This will open a few interesting doors.

**Comment by cfriedt** [comment]
* reformat `tests/kernel/threads/dynamic_thread_stack/testcase.yaml` in whatever way `yamllint` is cool with 🤷‍♂️

**Review by keith-packard — COMMENTED** [review]
I think you should always allocate stacks from the heap; reduces code complexity, avoids issues with the pre-allocated stacks being the wrong size and doesn't consume any more memory.

**Comment by cfriedt** [comment]
> I think you should always allocate stacks from the heap; reduces code complexity, avoids issues with the pre-allocated stacks being the wrong size and doesn't consume any more memory.

@keith-packard - I think that makes sense for a large majority of use cases, in particular userspace code, and we actually can get that behaviour via Kconfig (by specifying the pool size to be zero and enabling heap allocation).

The pool option is really there to support high reliability / safety critical _if they choose to use it_.

**Comment by andyross** [comment]
> The pool option is really there to support high reliability / safety critical if they choose to use it.

Even fragmentation is an issue in an RTOS.  Stacks are big, memories are small, uptimes are (sometimes very) long.  A dynamic spawned thread that works fine in testing in legacy code might suddenly die after two years in the field.  A core Zephyr paradigm is that the system heap should be 100% optional.  Everything that can be should be allocatable statically.

I'm actually dealing with a bunch of issues like this currently while fuzzing SOF, which is very heap-centric.  The allocation errors can be handled correctly (and are), but the resulting runtime failures are still failures, and prevent reaching further states.  The impact there is actually paradoxical: the random input tends to leak a lot of allocations and run out of space quickly, so in fact fuzz coverage drops like a rock after the first few thousand commands.  We're seeing failures come back from the (extremely parallel) oss-fuzz rig that I never saw over days and days of running locally.

**Comment by keith-packard** [comment]
> yep ... One thing though that I am thinking now is that having two separated API is probably better. 

Oh, so just exposing the `stack_alloc_pool` and `stack_alloc_dyn` apis directly instead of wrapping them in another function. That avoids problems with applications expecting pool allocations and getting surprised with dyn stacks.

One more request -- can we have the stack automatically freed when the thread exits?

**Comment by cfriedt** [comment]
> > yep ... One thing though that I am thinking now is that having two separated API is probably better.

> Oh, so just exposing the `stack_alloc_pool` and `stack_alloc_dyn` apis directly instead of wrapping them in another function. That avoids problems with applications expecting pool allocations and getting surprised with dyn stacks.

If that's the general consensus, it would probably be better to not even expose pool allocation this way - just have a separate K_THREAD_POOL() that only ever allocates at build time.

It would also make the pthread use case more complicated, which I was hoping to avoid, since that was the entire reason for doing this in a way to silently support both safety and non-safety in the same way.

@romkell, @andyross - do you have any preferences?

> One more request -- can we have the stack automatically freed when the thread exits?

That was tried a long time ago in a couple of different ways and it tended to be problematic and a bit racey.

Additionally, there was a vito from the safety side, because invariably then the malloc-ing code would be intermixed with the non-mallocing code.

**Comment by romkell** [comment]
> @romkell, @andyross - do you have any preferences?

My major point concerning safety is that we retain the API where we instantiate every thing statically on the appl. side (and pass to init function etc.).
To have a stack pool to separate the memory for stacks from other heap based mem allocation may bring some advantages. When exposing both APIs for pool and heap for the threads, the appl. developer can do any kind of mixed (pool and heap) dynamic stack allocation in the appl. Do we want to support that' When hiding underneath one API function the application developer decides whether to use the pool or heap approach via KConfig. What is the use case to allow both pool and heap stacks in one application?
For simplicity I would prefer one API, but can perfectly live with both exposed, and leave responsibility to the appl. developer.

**Comment by cfriedt** [comment]
> yep ... One thing though that I am thinking now is that having two separated API is probably better. The reason is that whe you are allocating from the pool the stack size is defined at build time by CONFIG_DYNAMIC_THREAD_STACK_SIZE (so the given parameter is not really needed and is confusing) while allocating from the heap we don't have this constraint. Also, K_USER does not make sense IMHO, since if you have userspace enabled they stacks declared in the pool are already user capable.

Technically, that is a maximum size for pooled stacks (maybe the Kconfig could be named better). The size parameter can still provide non-redundant information when size < max (especially in the case where an MMU or Stack Canaries are present).

In any case though, given that both the pure-heap, pure-pool, and mixed cases are realizable with a single API, I think it would eliminate _a non-trivial amount_ of code duplication and maybe some future bugs with virtually zero overhead (since kconfig-controlled code paths are dead-stripped by the linker in the current implementation).

To me that's a big selling feature that makes supporting the standard threading model in C, C++, pthreads (not to mention other languages) _dramatically_ easier.

Just at a glance, it does not seem possible to support statically allocated C11 or C++11 thread stacks at all, externally to the implementation.

https://en.cppreference.com/w/c/thread

https://en.cppreference.com/w/cpp/thread/thread/thread

I think it may be possible in Rust to specify a statically allocated thread stack.

So there is a significant probability we would be forced to duplicate all of that heap / pool logic for several language runtimes.

**Comment by ceolin** [comment]
> that the system heap should be 100% optional. Everything that can be should be allocatable statically.

I think that was well ... just the `K_USER` that option that I was talking about to be clear. Particularly because of the different behavior that this can lead on heap/pool allocation. Though we can choose the preferred way, if it fails the fallback is using the other one.

**Review by andyross — APPROVED** [review]
Should drop a +1 to move this along.  Really I think this is very reasonable.  It's a big area with lots of edges (especially once we start talking about userspace!), and no API is going to meet everyone's needs.  This seems straightforward and useful enough for common cases, and apps with complicated needs can use it as a reference if necessary.

File: kernel/Kconfig [hunk 9]
```diff
@@ -203,6 +203,68 @@ config THREAD_USERSPACE_LOCAL_DATA
...
+	select DYNAMIC_OBJECTS if USERSPACE
+	help
+	  Enable support for dynamic threads and stacks.
+
+if DYNAMIC_THREAD
+
+config DYNAMIC_THREAD_STACK_SIZE
+	int "Size of each pre-allocated thread stack"
+	default 1024 if !64BIT
+	default 2048 if 64BIT
```
Comment (dcpleung) [inline kernel/Kconfig:219]: `default MMU_PAGE_SIZE if MMU` as stack needs to use the full page for access control (especially for userspace).

Comment (ceolin) [reply kernel/Kconfig:219]: userspace is coming in my pr ... do not worry about it now please.

**Comment by cfriedt** [comment]
> Mentioned offline as well, but it would likely make sense to support flags for pool or heap as well

So if it's desirable, `k_thread_stack_alloc()` could accept `flags` like

`K_THREAD_STACK_FROM_HEAP`, or `K_THREAD_STACK_FROM_POOL`.

to indicate a suitable allocation strategy as well (a default of zero of these could imply either).

**Comment by cfriedt** [comment]
> Should drop a +1 to move this along.  Really I think this is very reasonable.  It's a big area with 

Amen to that. Lets get Flavio's changes in as well, so people can mix things up and develop organically.

Better to do this kind of thing early in the release cycle.

**Comment by ceolin** [comment]
> > Mentioned offline as well, but it would likely make sense to support flags for pool or heap as well
> 
> So if it's desirable, `k_thread_stack_alloc()` could accept `flags` like
> 
> `K_THREAD_STACK_FROM_HEAP`, or `K_THREAD_STACK_FROM_POOL`.
> 
> to indicate a suitable allocation strategy as well (a default of zero could imply either).

That is what I was thinking ... this way we can remove the build symbols about the allocation preferences and remove all fallbacks ... that is way is clear for the user from where he is getting the memory. Note that is not a blocker for me, we can do it incrementally.

**Review by ceolin — APPROVED** [review]
